src/server/page_server.cpp

```diff
@@ -107,64 +109,151 @@ page_server::connection_handler::receive_log_prior_list (tran_server_conn_t::seq
   log_Gl.get_log_prior_receiver ().push_message (std::move (a_ip.pull_payload ()));
 }
 
+template<class F>
 void
-page_server::connection_handler::receive_log_page_fetch (tran_server_conn_t::sequenced_payload &a_sp)
+page_server::connection_handler::push_async_response (F &&a_func, tran_server_conn_t::sequenced_payload &&a_sp)
+{
+  m_ps.get_responder ().async_execute (std::ref (*m_conn), std::move (a_sp),
+				       std::bind (std::forward<F> (a_func), this,
+					   std::placeholders::_1, std::placeholders::_2));
+}
+
+void
+page_server::connection_handler::fetch_log_page (cubthread::entry &context, std::string &a_payload)
 {
-  LOG_PAGEID pageid;
-  std::string message = a_sp.pull_payload ();
+  // Unpack the message data
+  LOG_PAGEID log_pageid;
+  assert (a_payload.size () == sizeof (log_pageid));
+  std::memcpy (&log_pageid, a_payload.c_str (), sizeof (log_pageid));
+
+  log_lsa fetch_lsa { log_pageid, 0 };
+  log_reader lr { LOG_CS_SAFE_READER };
 
-  assert (message.size () == sizeof (pageid));
-  std::memcpy (&pageid, message.c_str (), sizeof (pageid));
+  if (log_pageid == LOGPB_HEADER_PAGE_ID)
+    {
+      // Make sure log page header is updated
+      logpb_force_flush_header_and_pages (&context);
+    }
+
+  int error = lr.set_lsa_and_fetch_page (fetch_lsa);
 
+  // Response message
   if (prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE))
     {
-      _er_log_debug (ARG_FILE_LINE, "Received request for log from Transaction Server. Page ID: %lld \n", pageid);
+      _er_log_debug (ARG_FILE_LINE,
+		     "[READ LOG] Sending log page to Active Tran Server. Page ID: %lld Error code: %ld\n",
+		     log_pageid, error);
     }
 
-  auto callback_func = [this, sp = a_sp] (const LOG_PAGE *log_page, int error_code) mutable
-  {
-    on_log_page_read_result (std::move (sp), log_page, error_code);
-  };
+  // pack error first
+  a_payload = { reinterpret_cast<const char *> (&error), sizeof (error) };
 
-  m_ps.get_page_fetcher ().fetch_log_page (pageid, std::move (callback_func));
+  if (error == NO_ERROR)
+    {
+      // pack page data too
+      a_payload.append (reinterpret_cast<const char *> (lr.get_page ()), db_log_page_size ());
+    }
 }
 
 void
-page_server::connection_handler::receive_data_page_fetch (tran_server_conn_t::sequenced_payload &a_sp)
+page_server::connection_handler::receive_log_page_fetch (tran_server_conn_t::sequenced_payload &a_sp)
+{
+  push_async_response (&connection_handler::fetch_log_page, std::move (a_sp));
+}
+
+void
+page_server::connection_handler::fetch_data_page (cubthread::entry &context, std::string &a_payload)
```

Comment:
same here; move to `page_buffer`; `m_ps.get_replicator` can be replaced by the call to he global variable